### ISSUE TITLE
Virtual viewports and separate threads have undefined/odd behavior

### ISSUE THREAD

**Issue by Nukem9** [ocornut/imgui#1951]
**Version/Branch of Dear ImGui:**
1.63 WIP Viewport (https://github.com/ocornut/imgui/commit/23b67e6ff5b1b68d18d467dee8d0e3c8cd438c79)

**Back-end file/Renderer/OS:**
example_win32_directx11/D3D11/Windows 8.1

**Standalone, minimal, complete and verifiable example:**
Full copy+paste code for main.cpp https://gist.github.com/Nukem9/f206b9269b5d02263dc0520d85501293

Scenario:
The **main thread** does rendering and presentation (`ImGui_ImplDX11_RenderDrawData(ImGui::GetDrawData());` `g_pSwapChain->Present(1, 0);`)
A **secondary thread** handles window creation and the message loop.

I create the main program window + message loop on a separate thread first, then run the imgui/D3D11 render loop on the main thread. Everything seems to be OK if virtual viewports are disabled. If viewports are enabled and you drag a window outside the main window (2+ HWND's), all input stops working. Can't minimize, type, or move anything.

*One* of the issues is in [imgui_impl_win32.cpp](https://github.com/ocornut/imgui/blob/viewport/examples/imgui_impl_win32.cpp#L163) - GetActiveWindow(): `Retrieves the window handle to the active window attached to the **calling thread's** message queue.` Replacing it with GetForegroundWindow() only fixes the cursor position.

There's a messy solution (read: hack) to fix it, but wasn't sure if I should bother debugging it more. I'm not sure if the bug matters given how much of an edge case this is.

**Comment by ocornut** [comment]
I'm interested in digging and investigating problems happened in real world use but here I don't understand how the setup can work.

I'm not sure I understand it all, but for a start I don't see how the setup would be thread safe: you are dispatching messages leading to WndProc calling `ImGui_ImplWin32_WndProcHandler` while the other thread is running at maximum speed, so that message handler would directly poking inside imgui structures while imgui functions are called.

The code that calls `GetActiveWindow()` is a bit of an oddity because it is actually not really needed for the viewport branch, as we can just copy the raw win32 GetCursorPos() output to io.MousePos. That is to say, you can rework this in various way. (I think I will rework this code later and have explicit Viewport vs non-Viewport branches).

**Comment by Nukem9** [comment]
> I'm interested in digging and investigating problems happened in real world use but here I don't understand how the setup can work.

I'd put "real world use" very lightly here. I added a debugging overlay to Skyrim through reverse engineering and decided to try out the viewport branch yesterday. The weird threading setup is a result of that....please don't ~~waste~~ spend too much time on this issue.

> I'm not sure I understand it all, but for a start I don't see how the setup would be thread safe: you are dispatching messages leading to WndProc calling ImGui_ImplWin32_WndProcHandler while the other thread is running at maximum speed, so that message handler would directly poking inside imgui structures while imgui functions are called.

The non-viewport `ImGui_XXX_WndProcHandler` was lightweight enough that I used atomic operations/a lock. I know WndProcHandler changed for the viewport branch, but I'm willing to accept the risk for now. Even at 1700 FPS I haven't noticed any bugs *yet*. I omitted any synchronization code in the sample for simplicity.


Like I said, this is a rare use case. Thanks for the viewport branch though, I love using it!

**Comment by ocornut** [comment]
I’m not sure how you added the locks? The correct way here is to not change any of the io. field between NewFrame and EndFrame... your wndproc handler should probably record data somewhere else and you submit then yourself to the actual imgui data structures before NewFrame().

In order word, it is probably tricky to use much of the existing imgui_impl_win32 code as is.

I agree it’s an unusual situation. If you end up digging further don’t hesitate to write about what you find. I wouldn’t find making small changes to the backend if they can facilitate more situations.

**Comment by Nukem9** [comment]
I experimented a bit more but in the end I couldn't find an seamless way to handle this. It's "possible" as long as I add some code in impl_win32.cpp that allows a ping-pong between the 2 threads. Regardless, my hacky solution still works for me.
If you're wondering why - `ImGui::UpdatePlatformWindows` calls `ImGui_ImplWin32_CreateWindow` which calls `CreateWindowEx`, binding the window to the main rendering thread. There's no getting around it.

Since you mentioned `I wouldn’t find making small changes to the backend if they can facilitate more situations`:
- Using `GetActiveWindow` is a poor choice in my opinion. The return value is exclusive to the calling thread.
- [DestroyWindow](https://github.com/ocornut/imgui/blob/viewport/examples/imgui_impl_win32.cpp#L466) also can't touch windows on different threads. I replaced it with `::PostMessage(data->Hwnd, WM_CLOSE, 0, 0);`

Note: Your code is 100% fine as-is. They're only suggestions.
